File: testPointer.c
/* 
 * File:   testPointer.c
 * Author: tibo
 *
 * Created on 4 December 2014, 11:37 PM
 */

#include <stdio.h>
#include <stdlib.h>

struct animal{
   void (*addOne)(struct animal* self);
   int (*getCount)(struct animal* self);
};

struct dog{
   struct animal animal;
   int count;
} ;
void dogAdd(struct animal* self) {
    ((struct dog*)self)->count++;
}

int dogCount(struct animal* self) {
    return ((struct dog*)self)->count;
}

struct cat{
   struct animal animal;
   int count;
} ;
void catAdd(struct animal* self) {
    ((struct cat*)self)->count++;
}

int catCount(struct animal* self) {
    return ((struct cat*)self)->count;
}

typedef struct zebra{
   struct animal animal;
   int count;
} Zebra;
void zebraAdd(struct animal* self) {
    ((struct zebra*)self)->count++;
}

int zebraCount(struct animal* self) {
    return ((struct zebra*)self)->count;
}

int main(int argc, char** argv) {
    struct dog myDog = {
        .animal={
            .addOne=&dogAdd,
            .getCount=&dogCount
        },
        .count=0
    };
    struct dog myCat = {
        .animal={
            .addOne=&catAdd,
            .getCount=&catCount
        },
        .count=0
    };
    struct dog myZebra = {
        .animal={
            .addOne=&zebraAdd,
            .getCount=&zebraCount
        },
        .count=0
    };
    struct animal* myAnimals[3] ={(struct animal*)&myDog,(struct animal*)&myCat,(struct animal*)&myZebra};
    int i;
    for (i = 0; i < 10000000; i++) {
        int j;
            struct animal* a = myAnimals[i%3];
            a->addOne(a);
        
    }
    int j;
    for (j = 0; j < 3; j++) {
        struct animal* a = myAnimals[j];
        printf("count : %d",a->getCount(a));
    }

    return (EXIT_SUCCESS);
}

File: TestPointer.java
import java.util.regex.Matcher;
import java.util.regex.Pattern;

interface Animal{
    void addOne();
    int getTotal();
}

class Dog implements Animal{

    static int count = 0;

    @Override
    public void addOne() {
        count++;
    }

    @Override
    public int getTotal() {
        return count;
    }
}
class Cat implements Animal{

    static int count = 0;

    @Override
    public void addOne() {
        count++;
    }

    @Override
    public int getTotal() {
        return count;
    }
}

class Zebra implements Animal{

    static int count = 0;

    @Override
    public void addOne() {
        count++;
    }

    @Override
    public int getTotal() {
        return count;
    }
}
public class TestPointer {


    public static void main(String[] args) {
        Animal[] myAnimals = new Animal[]{new Dog(),new Cat(),new Zebra()};
        doIt(myAnimals);
        doIt(myAnimals);

    }

    private static void doIt(Animal[] myAnimals) {
        long start = System.currentTimeMillis();
        for (int i = 0; i < 10_000_000 ; i++) {
            for (Animal myAnimal : myAnimals) {
                myAnimal.addOne();
            }
        }
        long end = System.currentTimeMillis();
        System.out.println("time = " + (end-start));
    }
}

